### PR TITLE
Refactor WagedRebalancer and add comments

### PR DESCRIPTION
**Pull request by qqu0127**
Create standalone classes for partial and global rebalance to make the class more modular and easier to manage.

### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

Resolves https://github.com/apache/helix/issues/2430

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

Refactor `WagedRebalancer` by separating out components to their standalone classes.
Currently the WagedRebalancer class is complicated and contains multiple processes. With everything in one class, it's hard to read and even worse to manage complexity. 
Creating their classes for partial and global rebanace have a clear cut on responsibility, and make it possible for future refactoring toward microservices.

This PR does the following:
1. Move partial and global rebalance process from WagedRebalancer to their own classes
2. Remove process-specific member variables and methods from WagedRebalancer
3. Add clarifying comments for partial and global rebalance.

No new logic is introduced, this PR is pure refactoring.

### Tests

- [X] The following tests are written for this issue:

Existing tests in `TestWagedRebalancer`

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

Test for `TestWagedRebalancer`
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.535 s - in org.apache.helix.controller.rebalancer.waged.TestWagedRebalancer
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.6:report (generate-code-coverage-report) @ helix-core ---
[INFO] Loading execution data file /Users/qqu/workspace/qqu-helix/helix-core/target/jacoco.exec
[INFO] Analyzed bundle 'Apache Helix :: Core' with 936 classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  40.842 s
[INFO] Finished at: 2023-04-03T15:42:46-04:00
[INFO] ------------------------------------------------------------------------

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Review by junkaixue — APPROVED** [review]
lgtm. Good job!

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/GlobalRebalanceRunner.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+
+/**
+ * Global Rebalance does the baseline recalculation when certain changes happen.
+ * The Global Baseline calculation does not consider any temporary status, such as participants' offline/disabled.
+ * Baseline is used as an anchor for {@link PartialRebalanceRunner}. Its computation takes previous baseline as an input.
+ * The Baseline is NOT directly propagated to the final output. It is consumed by the {link PartialRebalanceRunner}
+ * as an important parameter.
+ */
+class GlobalRebalanceRunner implements AutoCloseable {
```
Comment (desaikomal) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/GlobalRebalanceRunner.java:55]: nit: do we call this GlobalRebalance or BaseLine ?? I am fine with either but want to make sure we consistently use that term.

**Review by desaikomal — COMMENTED** [review]
This is extremely good change. Splitting the functionality is very useful.
It is not possible to look through each and every line of change as the change is really big for human brain to perform. 
Hopefully lift-n-shift worked. AFAIK you didn't fundamentally changed any behavior. just code  alignments. 
Amazingly cool

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/GlobalRebalanceRunner.java (hunk 1 above)
Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/GlobalRebalanceRunner.java:55]: The code always goes with global-partial. Baseline is what it computes. Let's keep it this way.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/AssignmentManager.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.HelixRebalanceException;
+import org.apache.helix.controller.stages.CurrentStateOutput;
+import org.apache.helix.model.ResourceAssignment;
+import org.apache.helix.monitoring.metrics.model.LatencyMetric;
+
+
```
[the remark this answers is not in the document]
Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/AssignmentManager.java:31]: Thanks, updated.

**Comment by qqu0127** [comment]
This PR is ready to merge, approved by @junkaixue 
Please use the original commit message:

Refactor WagedRebalancer and add comments
Create standalone classes for partial and global rebalance to make the class more modular and easier to manage.

**Review by rahulrane50 — APPROVED** [review]
This looks very clean! Good job Quincy! I just wish github has"moved code block" feature. it would have been very easy to make sure we don't miss anything in refactoring :)

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/AssignmentManager.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  /**
+   * @param assignmentMetadataStore
+   * @param currentStateOutput
+   * @param resources
+   * @return The current best possible assignment. If record does not exist in the
+   *         assignmentMetadataStore, return the current state assignment.
+   * @throws HelixRebalanceException
+   */
+  public Map<String, ResourceAssignment> getBestPossibleAssignment(
```
[the remark this answers is not in the document]
Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/AssignmentManager.java:82]: Thanks for the review.
That's a good idea should be doable, but let's leave that for future refactoring as this PR is big enough.

**Comment by xyuanlu** [comment]
Thanks for making the effort on code clean and refactor!

**Pull request merged**